File: Dynamic Programming/unique-paths-ii.py
from typing import List
# 120ms 5.48%
# 28.7MB 5.11%
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        m = len(obstacleGrid)
        n = len(obstacleGrid[0])
        d = [[0 for j in range(n)] for i in range(m)]
        for i in range(m):
            for j in range(n):
                if i == j == 0  and obstacleGrid[i][j] == 0:
                    d[i][j] = 1
                elif obstacleGrid[i][j] == 1:
                    d[i][j] = 0
                else:
                    d[i][j] = d[i-1][j] + d[i][j-1]
        return d[-1][-1]

a = Solution()
l = [
  [0]
]
print(a.uniquePathsWithObstacles(l))

File: Dynamic Programming/unique-paths-ii.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        int m=obstacleGrid.size(), n=obstacleGrid[0].size();
        vector<vector<int>> dp(m, vector<int>(n, 0));
        dp[0][0] = obstacleGrid[0][0] == 0 ? 1 : 0;
        for (int i=0; i<m; i++)
            for (int j=0; j<n; j++){
                if (i>0) dp[i][j] = obstacleGrid[i][j] == 0 ? (dp[i-1][j] + dp[i][j]) : 0;
                if (j>0) dp[i][j] = obstacleGrid[i][j] == 0 ? (dp[i][j-1] + dp[i][j]) : 0;
            }
        return dp.back().back();
    }
};

int main(){
    Solution sol;
    vector<vector<int>> obstacle = {{1}};
    cout << sol.uniquePathsWithObstacles(obstacle) << endl;
    return 0;
}